### PR TITLE
Stop Viewer Refresh

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
Redesigns some panels in the stop viewer to align the design more with the metro itinerary UI. Blocked by #657, #662, https://github.com/opentripplanner/otp-ui/pull/451, https://github.com/opentripplanner/otp-ui/pull/474, https://github.com/opentripplanner/otp-ui/pull/491, https://github.com/opentripplanner/otp-ui/pull/480

#### Looking for feedback on:
The best way to transport the `hideTopBorder` info around. Rendering the top border when the background color is the same or similar to the route color looks very bad. 

The data-passing issue comes from the fact that the main stop viewer rows are being rendered differently but similarly to the related stop rows. The real solution is to rework the entire stop viewer, which is happening. The fact that most of this code will be scrapped soon anyway is why this `hideTopBorder` prop is sort of being attached in awkward places. If there is a quick solution to improve this I'd welcome it, but a larger refactoring should probably be saved for when the stop viewer is migrated to a "nearby" viewer.

### CODE REVIEW

File: lib/components/viewers/related-stops-panel.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/related-stops-panel.js:116]: Is there a reason why next arrivals in this panel are separate, whereas those for that stop are "glued" together?

Comment (miles-grant-ibigroup) [reply lib/components/viewers/related-stops-panel.js:116]: Aesthetics. If you glue the small ones together things look very busy if there are multiple colors going on and very messy if all lines are the same color. With the larger stop blocks we can rely on the visual hierarchy to help out more but in the small list we're only working with one horizontal row so things get muddy fast.

File: lib/components/viewers/stop-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-viewer.js:170]: Would love to see a mode icon in the header the same way it is shown in related stops!

Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-viewer.js:170]: I think we agreed to save this for the nearby view. I like it though -- great idea!

File: lib/components/viewers/stop-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-viewer.js:176]: The font color should be set where the ambient background color is set (multiple instances).

File: lib/components/viewers/stop-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-viewer.js:172]: Surround the `||` statement with parentheses (try with this stop http://localhost:9966/#/stop/2:841 to see the difference).

File: lib/components/viewers/stop-viewer.js
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-viewer.js:192]: Nit: could that go inside the span?

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:45]: Might need to set the color of the user agent calendar widget too.

File: lib/components/viewers/pattern-row.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/pattern-row.tsx:91]: I don't see that applied to the `RAPID` bus (try with http://localhost:9966/#/stop/2:841)

Comment (binh-dam-ibigroup) [reply lib/components/viewers/pattern-row.tsx:91]: Is this going to accommodate long long `routeShortName` that some agencies on the West Coast use?

File: lib/components/viewers/pattern-row.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/pattern-row.tsx:109]: Sounds like you need another RouteRenderer component.

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:102]: Sort CSS (multiple instances)

File: lib/util/viewer.js
Comment (binh-dam-ibigroup) [inline lib/util/viewer.js:398]: I would much rather get-a-`RouteNameRenderer`-based-on-settings, and each route renderer can decide which colors/shapes/etc to use based on the mode and other route attributes.

File: lib/util/viewer.js
Comment (binh-dam-ibigroup) [inline lib/util/viewer.js:438]: If you create different kinds of route renderers, each can decide whether to render a border or not.

File: lib/components/viewers/stop-time-cell.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-time-cell.tsx:103]: Add a `title` and `aria-label` to this element that say the real-time status.

File: lib/components/viewers/stop-time-cell.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-time-cell.tsx:108]: Move the `realtime` portion to the `PulsingRss` component and move the other styles to a styled wrapper?

File: lib/components/viewers/related-stops-panel.js
Comment (binh-dam-ibigroup) [reply lib/components/viewers/related-stops-panel.js:116]: I still think "gluing" the arrivals by stop would provide visual separation between nearby stops. The headings don't outline groups of arrivals as nicely.

Comment (miles-grant-ibigroup) [reply lib/components/viewers/related-stops-panel.js:116]: I think the real solution is to move to the nearby view we've been discussing for some time. The entire idea of the "nearby stops" view below the stop has never sat right with me.

File: lib/components/narrative/metro/default-route-renderer.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/default-route-renderer.tsx:25]: It's definitely a hack... The issue comes from the fact that the `RouteRenderer` component doesn't really exist if that makes sense. Since it's pulled in from context, we can't change the props. I could try to find the exact detailed reason why this was done but it was a while ago and I remember I spent a good amount of time trying to figure out how to avoid it.

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:89]: We replace it here because the default one looks far too busy

File: lib/components/viewers/stop-schedule-table.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-schedule-table.tsx:115]: I originally had this, but it didn't color enough of the elements.

File: lib/components/viewers/stop-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-viewer.js:176]: I tried to do this, but there were too many instances of the wrong elements being colored.

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:109]: I'm not sure I understand what you mean. This RouteRenderer will be whichever one is defined in the config

File: lib/util/viewer.js
Comment (miles-grant-ibigroup) [reply lib/util/viewer.js:398]: I'm not sure if I understand what you mean. This is what's happening now. The reason that some of the syntax of this is so strange is precisely because the custom route renderer can generate any HTML it would like.

File: lib/util/viewer.js
Comment (miles-grant-ibigroup) [reply lib/util/viewer.js:438]: I think I see what you mean now. But wouldn't this wrapper route renderer still wind up passing the same prop to the underlying route renderer? I agree this prop is poorly named and should instead be used to send context information to the route renderer. Maybe something like `isBeingRenderedInCompactSpace` or similar.

Comment (binh-dam-ibigroup) [reply lib/util/viewer.js:438]: In my mind, each combination of visual styles (borders, how color is applied, shape...) would be a different renderer.

Regarding adjusting route names to the available space, one technique would be to render text as SVG, then the SVG would be fitted in the space available. See for instance https://github.com/opentripplanner/otp-ui/blob/47302686e8bb53256d3214c1cf1d54db7ebdf754/packages/transit-vehicle-overlay/src/VehicleIcon.tsx#L30 in https://github.com/opentripplanner/otp-ui/pull/470.

Comment (binh-dam-ibigroup) [reply lib/util/viewer.js:438]: So if the config is so that route background colors should be used in `PatternRow`, then use one kind of renderer. If the config sets a say white background color to `PatternRow`, use another kind of renderer.

Comment (miles-grant-ibigroup) [reply lib/util/viewer.js:438]: Also for the SVG trick -- are we sure that this doesn't have a11y implications?

Comment (binh-dam-ibigroup) [reply lib/util/viewer.js:438]: Correct, an 1`aria-label` is needed if using SVG.

Comment (binh-dam-ibigroup) [reply lib/util/viewer.js:438]: Going back to route renderers, to me, the sample below features four different route name renderers.
image

File: lib/components/viewers/next-arrival-for-pattern.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/next-arrival-for-pattern.tsx:81]: I am seeing inconsistent colors with this stop: http://localhost:9966/#/stop/5:21097, where the text for route 140 can be both black and white on the same screen.

File: lib/components/util/operator-logo.tsx
Comment (miles-grant-ibigroup) [reply lib/components/util/operator-logo.tsx:18]: I've changed the translation string to "logo for {operatorName}". This feels like a cleaner solution that hiding elements for screen readers?

File: lib/components/viewers/amenities-panel.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/amenities-panel.tsx:79]: Those styles weren't even doing anything!

File: lib/components/viewers/next-arrival-for-pattern.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/next-arrival-for-pattern.tsx:81]: This is because the related stops endpoint includes `textColor` while the stops endpoint doesn't. Repairing this involves a lot of redux plumbing, but it should be fixed once we source all the stop info from one endpoint in the future.

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:112]: Screen Shot 2022-11-14 at 9 26 52 AM
It sure is

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:94]: If we did this the flex box would affect the headsign as well. We could accomplish the div removal by using grid but that would add a lot of complexity to what is otherwise a simple component (style-wise).

File: lib/components/viewers/stop-time-cell.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-time-cell.tsx:102]: This makes sense in theory, except it's now an `li` in one place and a `div` in another. Not sure if there's a non-hack way to do that well...

File: lib/components/viewers/stop-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-viewer.js:172]: Very good catch thank you!

File: lib/components/viewers/stop-viewer.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-viewer.js:192]: No, the padding looked awkward.

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:91]: Yes if you have a look at septa which has long `routeShortName`s on it's rail lines they fit nicely.

File: lib/components/viewers/stop-time-cell.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-time-cell.tsx:108]: I think this would make it more difficult to a reader to understand what's going on

**Comment by miles-grant-ibigroup** [comment]
Thanks for the careful review @binh-dam-ibigroup! All changes are addressed or responded to. I couldn't find any regressions based on the changes, but I would appreciate a double-check, especially the CSS had quite a few adjustments.

File: lib/components/viewers/stop-time-cell.tsx
Comment (binh-dam-ibigroup) [reply lib/components/viewers/stop-time-cell.tsx:109]: You can place spans inside list items and divs.

Comment (miles-grant-ibigroup) [reply lib/components/viewers/stop-time-cell.tsx:109]: But wouldn't this add extra nesting where it's not needed

File: lib/components/viewers/stop-time-cell.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/stop-time-cell.tsx:26]: Following up on https://github.com/opentripplanner/otp-react-redux/pull/661/files#r1018446380, insert the following and remove the corresponding inline style from the containing span (that style attribute actually has no effect, so maybe a transform is not needed at all):

```css
          transform: scaleX(-1);
```

File: lib/components/viewers/stop-time-cell.tsx
Comment (binh-dam-ibigroup) [reply lib/components/viewers/stop-time-cell.tsx:115]: See follow-up comment in `PulsingRss`, and move the non-transform style into a `styled(StyledIconWrapperTextAlign)`.

File: lib/components/viewers/amenities-panel.tsx
Comment (binh-dam-ibigroup) [reply lib/components/viewers/amenities-panel.tsx:129]: Can use `<br />` tags instead of `<divs>`, with the advantage that you can improve the text indentation using:
```css
.related-item {
    text-indent: -27px; /* 22px image width + 5px margin */
    padding-left: 27px;
}
```

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Some more comments, but the blocking item is regarding fitting not-so-long route names such as "RAPID" into the available in `PatternRow`.

File: i18n/es.yml
Comment (miles-grant-ibigroup) [reply i18n/es.yml:496]: I had to add the strings because of the i18n messages check lint step!

File: lib/components/narrative/metro/default-route-renderer.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/default-route-renderer.tsx:17]: I agree that these extra props are messy, but I don't think either of us have an ideal solution here.

I think having the route renderer accept a `leg` may have been a bit ambitious. A future refactor that might be good is to have the route renderer accept only route name and color. This should be enough to still do the custom styling we're currently doing, and it would make this `onColoredBackground` prop less messy.

File: lib/components/util/operator-logo.tsx
Comment (miles-grant-ibigroup) [reply lib/components/util/operator-logo.tsx:21]: That's a good aria label!

File: lib/components/viewers/RouteRow.js
Comment (miles-grant-ibigroup) [reply lib/components/viewers/RouteRow.js:153]: Slept right through that one, thanks!

File: lib/components/viewers/live-stop-times.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/live-stop-times.tsx:171]: It could be, but we'd then have to assemble it back together within the file for `next-arrival-for-pattern`. We can't ship operator as its own prop there since the related stops panel deals with a list of routes. Matching those arrays up would get messy.

File: lib/components/viewers/live-stop-times.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/live-stop-times.tsx:175]: sure!

File: lib/components/viewers/next-arrival-for-pattern.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/next-arrival-for-pattern.tsx:22]: See previous explanation -- I originally had it like this but found this solution to be cleaner than doing weird array length matching. I'll add a comment

File: lib/components/viewers/pattern-row.tsx
Comment (binh-dam-ibigroup) [reply lib/components/viewers/pattern-row.tsx:91]: This item still needs to be resolved.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
https://github.com/opentripplanner/otp-react-redux/pull/661#discussion_r1024492832 still needs to be resolved, I don't think that was updated in recent commits.

**Comment by miles-grant-ibigroup** [comment]
Thanks for catching that -- seems to be a blink issue. Will look into a polyfill. Screen Shot 2022-11-21 at 11 12 13 AM

File: lib/components/viewers/amenities-panel.tsx
Comment (binh-dam-ibigroup) [reply lib/components/viewers/amenities-panel.tsx]: Pretty soon, we will start deeming `<div>`s inside `<li>`s to be invalid markup.

File: lib/components/viewers/pattern-row.tsx
Comment (binh-dam-ibigroup) [inline lib/components/viewers/pattern-row.tsx:91]: This or similar logic also needs to be applied to the related stops rows (see screenshot)

Screenshot 2022-11-21 at 3 29 39 PM

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Almost... one more change is needed regarding the route font size for related stops.

File: lib/components/viewers/amenities-panel.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/amenities-panel.tsx]: This change was made in another pr based on pr review...

File: lib/components/viewers/pattern-row.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/pattern-row.tsx:91]: Spent a long time trying to get the grid alignment right while also doing dynamic text sizing, in the end decided to just un-grid it. 31f58ccb70b6b8e55470e28c9eb85f09a8139dc3

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:238]: max-width causes unnecessary clipping:

```suggestion
  flex-grow: 1;
  margin-right: 4px; /* or insert a non-shrinkable fixed-size element after. */
```

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [inline lib/components/viewers/viewers.css:253]: Inside a `flex` element you should use flex attributes (although in some future life this might have to become a table again)
```suggestion
  flex-shrink: 0;
```

**Review by binh-dam-ibigroup — APPROVED** [review]
LGTM. Thanks for the changes! Just a few visual tweaks.

File: lib/components/viewers/viewers.css
Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css:238]: I tried this originally but ran into the following issue on long destinations. Have tried all kinds of combos of flex shrinking and growing, and even grid but couldn't come up with something as clean as the max width that behaved appropriately. Screen Shot 2022-11-28 at 8 54 10 AM

File: lib/components/viewers/viewers.css
Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css:253]: This one works, thanks!

**Comment by miles-grant-ibigroup** [comment]
Tests are failing pending `core-utils` updates being published and used in this PR

**Comment by miles-grant-ibigroup** [comment]
@daniel-heppner-ibigroup all dependencies are now working out of the box and this is ready for your review

**Comment by daniel-heppner-ibigroup** [comment]
I am seeing an issue where the stop times table headers are overlapping with the table contents. 
image

**Comment by binh-dam-ibigroup** [comment]
> I am seeing an issue where the stop times table headers are overlapping with the table contents. image

I thought I addressed that with #695 + the appropriate CSS file.

**Comment by miles-grant-ibigroup** [comment]
I think the specific fix got lost in a merge... I will attempt to fix now

**Comment by miles-grant-ibigroup** [comment]
> > I am seeing an issue where the stop times table headers are overlapping with the table contents. image
> 
> I thought I addressed that with #695 + the appropriate CSS file.

Yes you are right the appropriate CSS file being key to this fix working. @daniel-heppner-ibigroup the CSS your config is using might need to be updated.

**Review by daniel-heppner-ibigroup — APPROVED** [review]
Thank you for the great work on this! It looks really great! Had a few comments and suggestions, I think I can still approve though.

File: lib/components/viewers/viewers.css
Comment (binh-dam-ibigroup) [reply lib/components/viewers/viewers.css:241]: For this comment, I still think replacing `max-width` with `margin-right` renders better on wide screens.

Comment (miles-grant-ibigroup) [reply lib/components/viewers/viewers.css:241]: I agree, but this specific reflow issue to me is unacceptable

File: lib/components/viewers/amenities-panel.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/amenities-panel.tsx:93]: This is really great thanks so much!

File: lib/components/narrative/metro/default-route-renderer.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/default-route-renderer.tsx:7]: Well spotted thanks

File: lib/components/narrative/metro/default-route-renderer.tsx
Comment (miles-grant-ibigroup) [reply lib/components/narrative/metro/default-route-renderer.tsx:37]: Good idea changed!

File: lib/components/viewers/amenities-panel.tsx
Comment (miles-grant-ibigroup) [reply lib/components/viewers/amenities-panel.tsx:6]: Good catch. Gone!

**Comment by miles-grant-ibigroup** [comment]
Thanks @daniel-heppner-ibigroup for those notes! This is now ready to merge once the Percy fix #730 is merged

**Pull request merged**
